File: src/types/report.ts
// ── Report 1: Daily Sales Summary (REPT-03) ──

export interface DailySalesRow {
  date: string;
  sale_count: number;
  item_count: number;
  gross_sales: number;
  discounts: number;
  tax_amount: number;
  net_sales: number;
  profit: number;
}

// ── Report 2: Monthly P&L (REPT-04) ──

export interface MonthlyPnLRow {
  month: string;
  sale_count: number;
  total_revenue: number;
  total_cogs: number;
  gross_profit: number;
  total_refunds: number;
  write_off_losses: number;
  net_profit: number;
}

// ── Report 3: Top Selling Medicines (REPT-05) ──

export interface TopSellerRow {
  medicine_id: number;
  medicine_name: string;
  generic_name: string | null;
  total_qty: number;
  total_revenue: number;
  total_profit: number;
}

// ── Report 4: Slow-Moving Stock (REPT-06) ──

export interface SlowMovingRow {
  medicine_id: number;
  medicine_name: string;
  category: string | null;
  current_stock: number;
  total_investment: number;
}

// ── Report 5: Low Stock (REPT-07) ──

export interface LowStockRow {
  medicine_id: number;
  medicine_name: string;
  category: string | null;
  reorder_level: number;
  current_stock: number;
  unit: string;
  deficit: number;
}

// ── Report 6: Expiry Report (REPT-08) ──

export interface ExpiryReportDetailRow {
  batch_id: number;
  medicine_id: number;
  medicine_name: string;
  batch_code: string | null;
  original_qty: number;
  remaining_qty: number;
  unit_cost: number;
  expiry_date: string;
  days_remaining: number;
  potential_loss: number;
  status: string;
}

// ── Report 7: Supplier Purchase History (REPT-09) ──

export interface SupplierPurchaseRow {
  supplier_id: number;
  company_name: string;
  purchase_count: number;
  item_count: number;
  total_spent: number;
  avg_order_value: number;
  last_purchase_date: string | null;
}

// ── Report 8: Sales by User (REPT-10) ──

export interface SalesByUserRow {
  user_id: number;
  full_name: string;
  role: string;
  sale_count: number;
  item_count: number;
  total_sales: number;
  total_profit: number;
  avg_profit_per_sale: number;
}

// ── Report 9: Profit Margin (REPT-11) ──

export interface ProfitMarginRow {
  medicine_id: number;
  medicine_name: string;
  category: string | null;
  times_sold: number;
  total_qty: number;
  avg_sell_price: number;
  avg_cost: number;
  avg_margin_per_unit: number;
  margin_pct: number;
  total_profit: number;
}

// ── Settings Write Payload ──

export interface UpdateSettingsPayload {
  pharmacy_name?: string | null;
  owner_name?: string | null;
  phone?: string | null;
  address?: string | null;
  logo_path?: string | null;
  owner_email?: string | null;
  default_tax_rate?: number | null;
  tax_enabled_default?: boolean | null;
  cashier_discount_enabled?: boolean | null;
  currency_symbol?: string | null;
  default_reorder_level?: number | null;
  expiry_warning_days?: number | null;
  expiry_critical_days?: number | null;
  auto_backup_time?: string | null;
  local_backup_path?: string | null;
}

// ── Backup DTOs ──

export interface BackupStatus {
  last_backup_time: string | null;
  last_backup_status: string | null;
  google_drive_connected: boolean;
}

export interface BackupResult {
  success: boolean;
  message: string;
  path: string;
}

export interface BackupFileInfo {
  id: string;
  name: string;
  created_time: string;
}

// ── Database Status ──

export interface DbStatus {
  db_path: string;
  db_size_bytes: number;
  migration_version: number;
  total_migrations: number;
  wal_mode: boolean;
  foreign_keys: boolean;
}

File: src-tauri/src/models/report.rs
use serde::{Deserialize, Serialize};

// ── Report 1: Daily Sales Summary (REPT-03) ──

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DailySalesRow {
    pub date: String,
    pub sale_count: i64,
    pub item_count: i64,
    pub gross_sales: f64,
    pub discounts: f64,
    pub tax_amount: f64,
    pub net_sales: f64,
    pub profit: f64,
}

// ── Report 2: Monthly P&L (REPT-04) ──

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MonthlyPnLRow {
    pub month: String,
    pub sale_count: i64,
    pub total_revenue: f64,
    pub total_cogs: f64,
    pub gross_profit: f64,
    pub total_refunds: f64,
    pub write_off_losses: f64,
    pub net_profit: f64,
}

// ── Report 3: Top Selling Medicines (REPT-05) ──

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TopSellerRow {
    pub medicine_id: i64,
    pub medicine_name: String,
    pub generic_name: Option<String>,
    pub total_qty: i64,
    pub total_revenue: f64,
    pub total_profit: f64,
}

// ── Report 4: Slow-Moving Stock (REPT-06) ──

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SlowMovingRow {
    pub medicine_id: i64,
    pub medicine_name: String,
    pub category: Option<String>,
    pub current_stock: i64,
    pub total_investment: f64,
}

// ── Report 5: Low Stock (REPT-07) ──

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct LowStockRow {
    pub medicine_id: i64,
    pub medicine_name: String,
    pub category: Option<String>,
    pub reorder_level: i64,
    pub current_stock: i64,
    pub unit: String,
    pub deficit: i64,
}

// ── Report 6: Expiry Report (REPT-08) ──
// Extended version with batch_code, potential_loss, status fields.
// Distinct from the simpler ExpiryReportRow in batch.rs.

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExpiryReportDetailRow {
    pub batch_id: i64,
    pub medicine_id: i64,
    pub medicine_name: String,
    pub batch_code: Option<String>,
    pub original_qty: i64,
    pub remaining_qty: i64,
    pub unit_cost: f64,
    pub expiry_date: String,
    pub days_remaining: i64,
    pub potential_loss: f64,
    pub status: String,
}

// ── Report 7: Supplier Purchase History (REPT-09) ──

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SupplierPurchaseRow {
    pub supplier_id: i64,
    pub company_name: String,
    pub purchase_count: i64,
    pub item_count: i64,
    pub total_spent: f64,
    pub avg_order_value: f64,
    pub last_purchase_date: Option<String>,
}

// ── Report 8: Sales by User (REPT-10) ──

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SalesByUserRow {
    pub user_id: i64,
    pub full_name: String,
    pub role: String,
    pub sale_count: i64,
    pub item_count: i64,
    pub total_sales: f64,
    pub total_profit: f64,
    pub avg_profit_per_sale: f64,
}

// ── Report 9: Profit Margin (REPT-11) ──

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProfitMarginRow {
    pub medicine_id: i64,
    pub medicine_name: String,
    pub category: Option<String>,
    pub times_sold: i64,
    pub total_qty: i64,
    pub avg_sell_price: f64,
    pub avg_cost: f64,
    pub avg_margin_per_unit: f64,
    pub margin_pct: f64,
    pub total_profit: f64,
}

// ── Settings Write Payload ──

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct UpdateSettingsPayload {
    // Pharmacy Info
    pub pharmacy_name: Option<String>,
    pub owner_name: Option<String>,
    pub phone: Option<String>,
    pub address: Option<String>,
    pub logo_path: Option<String>,
    pub owner_email: Option<String>,
    // Financial
    pub default_tax_rate: Option<f64>,
    pub tax_enabled_default: Option<bool>,
    pub cashier_discount_enabled: Option<bool>,
    pub currency_symbol: Option<String>,
    // Inventory
    pub default_reorder_level: Option<i64>,
    pub expiry_warning_days: Option<i64>,
    pub expiry_critical_days: Option<i64>,
    // Backup
    pub auto_backup_time: Option<String>,
    pub local_backup_path: Option<String>,
}

// ── Backup DTOs ──

#[derive(Debug, Clone, Serialize)]
pub struct BackupStatus {
    pub last_backup_time: Option<String>,
    pub last_backup_status: Option<String>,
    pub google_drive_connected: bool,
}

#[derive(Debug, Clone, Serialize)]
pub struct BackupResult {
    pub success: bool,
    pub message: String,
    pub path: String,
}

#[derive(Debug, Serialize)]
pub struct BackupFileInfo {
    pub id: String,
    pub name: String,
    pub created_time: String,
}
